src/openms/source/ANALYSIS/QUANTITATION/IsotopeLabelingMDVs.cpp

```diff
@@ -0,0 +1,327 @@
+      {
+        corrected_value += correction_matrix_eigen_inversed(i,j) * MDV_observed[j];
+      }
+      corrected_feature.getSubordinates().at(i).setIntensity(corrected_value);
+    }
+  }
+
+  void IsotopeLabelingMDVs::isotopicCorrections(
+    const FeatureMap& normalized_featureMap,
+    FeatureMap& corrected_featureMap,
+    const Matrix<double>& correction_matrix,
+    const std::string correction_matrix_agent)
+  {
+    for (const Feature& feature : normalized_featureMap)
+    {
+      Feature corrected_feature;
+      isotopicCorrection(feature, corrected_feature, correction_matrix, correction_matrix_agent);
+      corrected_featureMap.push_back(corrected_feature);
+    }
+  }
+
+  void IsotopeLabelingMDVs::calculateIsotopicPurity(
+    const Feature& normalized_featuremap,
+    Feature& featuremap_with_isotopic_purity,
+    std::vector<double>& experiment_data,
+    std::string& isotopic_purity_name)
+  {
+    featuremap_with_isotopic_purity = normalized_featuremap;
+    
+    if (!experiment_data.empty())
+    {
+      double experiment_data_peak = 0.0;
+      std::vector<double>::iterator max_it = std::max_element(experiment_data.begin(), experiment_data.end());
+      uint64_t experiment_data_peak_idx = std::distance(experiment_data.begin(), max_it);
+      experiment_data_peak = experiment_data[experiment_data_peak_idx];
+      
+      if (experiment_data_peak_idx >= 1 && experiment_data_peak != 0.0)
+      {
+        double previous_experiment_data_peak = experiment_data[experiment_data_peak_idx - 1];
+        double isotopic_purity = experiment_data_peak_idx / (experiment_data_peak_idx + (previous_experiment_data_peak / experiment_data_peak));
+        featuremap_with_isotopic_purity.setMetaValue(isotopic_purity_name, isotopic_purity);
+      }
+    }
+  }
+
+  void IsotopeLabelingMDVs::calculateIsotopicPurities(
+    const FeatureMap& normalized_featureMap,
+    FeatureMap& featureMap_with_isotopic_purity,
+    std::vector<double>& experiment_data,
+    std::string& isotopic_purity_name)
+  {
+    for (const Feature& feature : normalized_featureMap)
+    {
+      Feature feature_with_isotopic_purity;
+      calculateIsotopicPurity(feature, feature_with_isotopic_purity, experiment_data, isotopic_purity_name);
+      featureMap_with_isotopic_purity.push_back(feature_with_isotopic_purity);
+    }
+  }
+  
+  void IsotopeLabelingMDVs::calculateMDVAccuracy(
+    const Feature& normalized_feature,
+    Feature& feature_with_accuracy_info,
+    const std::vector<double>& fragment_isotopomer_measured,
+    const std::vector<double>& fragment_isotopomer_theoretical)
+  {
+    feature_with_accuracy_info = normalized_feature;
+    
+    std::vector<double> fragment_isotopomer_abs_diff;
+    for (size_t i = 0; i < fragment_isotopomer_theoretical.size(); ++i)
+    {
+      fragment_isotopomer_abs_diff.push_back(std::abs(fragment_isotopomer_theoretical[i] - fragment_isotopomer_measured[i]));
+    }
+    
+    double diff_mean = 0.0;
```

Comment:
some of these loops are not needed (since you can calculate the sum on the fly),

but you might also want to consider:
OpenMS::StatisticFunctions::MAD -- for median absolute deviate (instead of mean absolute deviation -- which should be susceptible to outliers)

Reply:
just a thought: if you really want the code to do exactly as it is, you might want to consider adding 'MeanAbsoluteDeviation' to `OpenMS::StatisticFunctions` and use it from there.

Reply:
we have Math::MAD() already

Reply:
> 
> 
> we have Math::MAD() already

Yep. That's what I meant with `OpenMS::StatisticFunctions::MAD`.
But we do not have MeanAbsoluteDeviation (which I'm not sure makes even sense)

Reply:
> > we have Math::MAD() already
> 
> Yep. That's what I meant with `OpenMS::StatisticFunctions::MAD`.
> But we do not have MeanAbsoluteDeviation (which I'm not sure makes even sense)

I have updated StatisticFunctions.h with `MeanAbsoluteDeviation`.